crates/ika-core/src/dwallet_mpc/mpc_session.rs

```diff
@@ -0,0 +1,505 @@
+        ))
+    }
+
+    /// Report that the session failed because of malicious actors.
+    /// Once a quorum of validators reports the same actor, it is considered malicious.
+    /// The session will be continued, and the malicious actors will be ignored.
+    fn new_dwallet_report_failed_session_with_malicious_actors(
+        &self,
+        report: MaliciousReport,
+    ) -> DwalletMPCResult<ConsensusTransaction> {
+        Ok(
+            ConsensusTransaction::new_dwallet_mpc_session_failed_with_malicious(
+                self.epoch_store()?.name,
+                report,
+            ),
+        )
+    }
+
+    /// Stores a message in the serialized messages map.
+    /// Every new message received for a session is stored.
+    /// When a threshold of messages is reached, the session advances.
+    pub(crate) fn store_message(&mut self, message: &DWalletMPCMessage) -> DwalletMPCResult<()> {
+        let source_party_id =
+            authority_name_to_party_id(&message.authority, &*self.epoch_store()?)?;
+
+        let current_round = self.serialized_messages.len();
+        match self.serialized_messages.get_mut(message.round_number) {
+            Some(party_to_msg) => {
+                if party_to_msg.contains_key(&source_party_id) {
+                    // Duplicate.
+                    // This should never happen, as the consensus uniqueness key contains only the origin authority,
+                    // session ID and MPC round.
+                    return Ok(());
+                }
+                party_to_msg.insert(source_party_id, message.message.clone());
+            }
+            // If next round.
+            None if message.round_number == current_round => {
+                let mut map = HashMap::new();
+                map.insert(source_party_id, message.message.clone());
+                self.serialized_messages.push(map);
+            }
+            None => {
+                // Unexpected round number; rounds should grow sequentially.
+                return Err(DwalletMPCError::MaliciousParties(vec![source_party_id]));
+            }
+        }
+        Ok(())
+    }
+
+    pub(crate) fn check_quorum_for_next_crypto_round(&self) -> ReadyToAdvanceCheckResult {
+        match self.status {
+            MPCSessionStatus::Active => {
+                if self.pending_quorum_for_highest_round_number == 0
+                    || self
+                        .weighted_threshold_access_structure
+                        .is_authorized_subset(
+                            &self
+                                .serialized_messages
+                                .get(self.pending_quorum_for_highest_round_number)
+                                .unwrap_or(&HashMap::new())
+                                .keys()
+                                .cloned()
+                                .collect::<HashSet<PartyID>>(),
+                        )
+                        .ok()
+                        .is_some()
```

Comment:
I'm not sure the conversion to `Option` is needed.

```suggestion
                        .is_ok()
```